src/Workspaces/Remote/Core/Diagnostics/DiagnosticComputer.cs

```diff
@@ -4,100 +4,141 @@
 
-        public DiagnosticComputer(Project project, DiagnosticAnalyzerInfoCache analyzerInfoCache)
+        public DiagnosticComputer(
+            DocumentId? documentId,
+            Project project,
+            TextSpan? span,
+            AnalysisKind? analysisKind,
+            DiagnosticAnalyzerInfoCache analyzerInfoCache)
         {
+            _document = documentId != null ? project.Solution.GetRequiredTextDocument(documentId) : null;
             _project = project;
+            _span = span;
+            _analysisKind = analysisKind;
             _analyzerInfoCache = analyzerInfoCache;
 
-            // we only track performance from primary branch. all forked branch we don't care such as preview.
+            // We only track performance from primary branch. All forked branch we don't care such as preview.
             _performanceTracker = project.IsFromPrimaryBranch() ? project.Solution.Workspace.Services.GetService<IPerformanceTrackerService>() : null;
         }
 
         public async Task<DiagnosticAnalysisResultMap<string, DiagnosticAnalysisResultBuilder>> GetDiagnosticsAsync(
             IEnumerable<string> analyzerIds,
             bool reportSuppressedDiagnostics,
-            bool logAnalyzerExecutionTime,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
             CancellationToken cancellationToken)
         {
-            var analyzerMap = CreateAnalyzerMap(_project);
-            var analyzers = GetAnalyzers(analyzerMap, analyzerIds);
+            var compilationWithAnalyzersData = await CompilationWithAnalyzersData.GetOrCreateAsync(
+                _project, isDocumentAnalysis: _document != null, cancellationToken).ConfigureAwait(false);
 
-            if (analyzers.Length == 0)
+            var analyzers = GetAnalyzers(compilationWithAnalyzersData.AnalyzerToIdMap, analyzerIds);
+            if (analyzers.IsEmpty)
             {
                 return DiagnosticAnalysisResultMap<string, DiagnosticAnalysisResultBuilder>.Empty;
             }
 
             var cacheService = _project.Solution.Workspace.Services.GetRequiredService<IProjectCacheService>();
             using var cache = cacheService.EnableCaching(_project.Id);
             var skippedAnalyzersInfo = _project.GetSkippedAnalyzersInfo(_analyzerInfoCache);
-            return await AnalyzeAsync(analyzerMap, analyzers, skippedAnalyzersInfo, reportSuppressedDiagnostics, logAnalyzerExecutionTime, cancellationToken).ConfigureAwait(false);
+            return await AnalyzeAsync(compilationWithAnalyzersData, analyzers, skippedAnalyzersInfo,
+                reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken).ConfigureAwait(false);
         }
 
         private async Task<DiagnosticAnalysisResultMap<string, DiagnosticAnalysisResultBuilder>> AnalyzeAsync(
-            BidirectionalMap<string, DiagnosticAnalyzer> analyzerMap,
+            CompilationWithAnalyzersData compilationWithAnalyzersData,
```

Comment:
>CompilationWithAnalyzersData compilationWithAnalyzersData [](start = 12, length = 57)

I'd pass the CompilationWithAnalyzers and id map as separate parameters, so that this method does not depend on these coming from the cache.